File: vb/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_9.vb
' This program is designed to convert hexadecimal numbers to decimal numbers.
' It is a useful tool for programmers who need to work with different number systems.
' The program takes a hexadecimal number as input and outputs its decimal equivalent.
' It demonstrates the process of parsing and converting number systems in Visual Basic.

Module HexToDecimalConverter

    ' Main entry point of the program
    Sub Main()
        ' Declare a variable to store the hexadecimal input from the user
        Dim hexInput As String

        ' Prompt the user to enter a hexadecimal number
        Console.WriteLine("Please enter a hexadecimal number:")
        hexInput = Console.ReadLine()

        ' Call the function to convert the hexadecimal number to decimal
        Dim decimalOutput As Integer = ConvertHexToDecimal(hexInput)

        ' Display the decimal equivalent of the entered hexadecimal number
        Console.WriteLine("The decimal equivalent of the hexadecimal number " & hexInput & " is " & decimalOutput)

        ' Wait for user input before closing the program
        Console.ReadLine()
    End Sub

    ' Function to convert a hexadecimal number to a decimal number
    Function ConvertHexToDecimal(ByVal hex As String) As Integer
        ' Initialize the decimal result to zero
        Dim decimalResult As Integer = 0

        ' Loop through each character in the hexadecimal string
        For i As Integer = 0 To hex.Length - 1
            ' Get the current character
            Dim currentChar As Char = hex(i)

            ' Determine the decimal value of the current hexadecimal digit
            Dim decimalValue As Integer
            If Char.IsDigit(currentChar) Then
                decimalValue = Asc(currentChar) - Asc("0")
            Else
                decimalValue = Asc(currentChar) - Asc("A") + 10
            End If

            ' Update the decimal result
            decimalResult = decimalResult * 16 + decimalValue
        Next

        ' Return the final decimal result
        Return decimalResult
    End Function

End Module

File: perl/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_9.pl
# This Perl program is designed to convert hexadecimal numbers to their decimal equivalents.
# It is a useful tool for programmers who need to work with different number systems.
# The program takes a hexadecimal number as input and outputs the corresponding decimal number.
# It is written in a verbose style with detailed comments to help understand each step of the process.

use strict;
use warnings;

# Function to convert a single hexadecimal digit to its decimal value
sub hex_digit_to_decimal {
    my $hex_digit = shift;
    my %hex_to_dec_map = (
        '0' => 0, '1' => 1, '2' => 2, '3' => 3, '4' => 4,
        '5' => 5, '6' => 6, '7' => 7, '8' => 8, '9' => 9,
        'A' => 10, 'B' => 11, 'C' => 12, 'D' => 13, 'E' => 14, 'F' => 15
    );
    return $hex_to_dec_map{uc($hex_digit)};
}

# Function to convert a hexadecimal string to a decimal number
sub hex_to_decimal {
    my $hex_string = shift;
    my $decimal_value = 0;
    my $length = length($hex_string);
    my $power = 0;

    # Iterate over each character in the hexadecimal string
    for (my $i = $length - 1; $i >= 0; $i--) {
        my $hex_digit = substr($hex_string, $i, 1);
        my $decimal_digit = hex_digit_to_decimal($hex_digit);
        $decimal_value += $decimal_digit * (16 ** $power);
        $power++;
    }

    return $decimal_value;
}

# Main program execution starts here
print "Enter a hexadecimal number: ";
my $hex_input = <STDIN>;
chomp($hex_input);

# Convert the input hexadecimal number to decimal
my $decimal_output = hex_to_decimal($hex_input);

# Print the resulting decimal number
print "The decimal equivalent of $hex_input is $decimal_output\n";

